File: DSAlgorithms/patterns/fast_slow_pointers/7_circular_arr_loop.cpp
#include<bits/stdc++.h>
class circular_arr_loop {
    public:
        static bool loop_exists(const std::vector<int> &arr)
        {
            for(int i = 0; i < arr.size(); i++)
            {
                bool is_forward = arr[i] >= 0;
                int slow = i;
                int fast = i;
                do{
                    slow = find_next_index(arr, is_forward, slow);
                    fast = find_next_index(arr, is_forward, fast);
                    if(fast != -1)
                        fast = find_next_index(arr, is_forward, fast);
                }
                while(slow != -1 && fast != -1 && slow != fast);
                if(slow != -1 && slow == fast)
                    return true;
            }
            return false;
        }
    private:
        static int find_next_index(const std::vector<int> &arr, bool is_forward, int current_index)
        {
            bool direction = arr[current_index] >= 0;
            if(is_forward != direction)
                return -1;
            int next_index = (current_index + arr[current_index] + arr.size()) % arr.size();
            if(next_index == current_index)
                next_index--;
            return next_index;
        }
};
int main(int argc, char *argv[])
{
    std::cout<<circular_arr_loop::loop_exists(std::vector<int>{1, 2, -1, 2, 2})<<'\n';
    std::cout<<circular_arr_loop::loop_exists(std::vector<int>{2, 2, -1, 2})<<'\n';
    std::cout<<circular_arr_loop::loop_exists(std::vector<int>{2, 2, -1, -2})<<'\n';
}

File: dsa_algo/patterns/fast_slow_pointers/7_circular_arr_loop.js
const circular_arr_loop = (arr)=> {
    for(let i = 0; i < arr.length; i++) 
    {
        let is_forward = arr[i] >= 0;
        let slow = i;
        let fast= i;
        while(true)
        {
            slow = find_next_index(arr, is_forward, slow);
            fast = find_next_index(arr, is_forward, fast);
            if(fast !== -1)
                fast = find_next_index(arr, is_forward, fast);
            if(slow === -1 || fast === -1 || slow === fast)
                break;
        }
        if(slow !== -1 && slow === fast)
            return true;
    }
    return false;
}

const find_next_index = (arr, is_forward, current_index) => {
    direction = arr[current_index] >= 0;
    if(is_forward !== direction)
        return -1;
    next_index = (current_index + arr[current_index]) % arr.length;
    if(next_index < 0)
        next_index += arr.length;
    if(next_index === current_index)
        next_index--;
    return next_index;
}

console.log(circular_arr_loop([1, 2, -1, 2, 2]));
console.log(circular_arr_loop([2, 2, -1, 2]));
console.log(circular_arr_loop([2, 2, -1, -2]));
